Add capsule migration option

wzzrd opened the pull request:
I added a --new-capsule option, as discussed on Rocket.Chat. 

It took some going back and forth and trying out various existing methods. I used some, like clean_katello_agent(), and ignored some others (clean_puppet()) for simplicity's sake.

I have tested with a RHEL7 machine, but I see no reason this shouldn't work on a RHEL6 client. The only piece of code that is not mine that I'm potentially hitting is the put_json() function, which seems to be called in disassociate_host(). Haven't tested, but I don't think it breaks anything.

Commit:
Allow put_json to actually receive jdata

Sometimes, put_json can work with just a URL, sometimes, it needs actual
json data.

Commit:
Add --new-capsule option

This option allows an administrator to use bootstrap.py to move a system
to another Capsule for content, Puppet and OpenSCAP using the
bootstrap.py script.

Some manual intervention still required because of missing API calls.
Will file BZs for those (revoke Puppet cert, create autosign entry).

Commit:
Refactor ugly json.loads lines

Created a function to DRY this out

Commit:
Fix Travis errors

evgeni commented on bootstrap.py:
```diff
@@ -879,7 +891,6 @@ class MEOptions:
                 import simplejson as json
             except ImportError:
                 print_error("Could not install python-simplejson")
-                sys.exit(1)
```
Why? simplejson is required on RHEL5.

wzzrd replied:
My bad. Removed this by accident :|

evgeni commented on bootstrap.py:
```diff
@@ -918,6 +929,44 @@ class MEOptions:
+        exec_failok("/sbin/service puppet stop")   # failok because people might be running Puppet from cron
+
+        print_running("Updating Puppet configuration")
+        exec_failexit("/usr/bin/puppet config set --section agent server %s" % options.foreman_fqdn)
```
this does not work on RHEL5, the puppet agent there does not support "puppet config set" :(

wzzrd replied:
Let me look into using ConfigParser a bit. Might fit this nicely. Again: don't want to nuke people's `/etc/puppet/puppet.conf` just to change a single settings :D

wzzrd replied:
It turns out SafeConfigParser is pretty picky on what it accepts as a valid config file. This is a no-go.  I might end up running `sed`instead, ffs :P

evgeni replied:
Yeah, I tried ConfigParser and it choked pretty badly on the default puppet.conf, so that is why we have this ugly blob to begin with.

wzzrd replied:
Ok, went with sed for now. Two actually, because stock RHEL5 puppet.conf has a ca_server *and* a server option.

evgeni commented on bootstrap.py:
```diff
@@ -918,6 +929,44 @@ class MEOptions:
+        # > API doesn't have a public provision for creating auto-sign entries yet!
+        print_running("Removing katello-*, gofer and installing new Katello certificate")
+        clean_katello_agent()
+        exec_failexit("rpm -Uvh http://%s/pub/katello-ca-consumer-latest.noarch.rpm" % options.foreman_fqdn)
```
This should be `get_bootstrap_rpm()`?

wzzrd replied:
Quite. This seems to be a remnant of my own refactoring. Will change, new pull request to follow

evgeni commented on bootstrap.py:
```diff
@@ -918,6 +929,44 @@ class MEOptions:
+        update_host_capsule_mapping(options.foreman_fqdn, API_PORT, "openscap_proxy_id", capsule_id, host_id)
+
+        print_running("Restarting goferd and rhsmcertd")
+        exec_failexit("/sbin/service rhsmcertd restart")
```
This should be `enable_rhsmcertd()`?

wzzrd replied:
Will change, new pull request to follow

evgeni commented on bootstrap.py:
```diff
@@ -918,6 +929,44 @@ class MEOptions:
+
+        print_running("Restarting goferd and rhsmcertd")
+        exec_failexit("/sbin/service rhsmcertd restart")
+        exec_failexit("/sbin/service goferd restart")
```
This is already done in `install_katello_agent()`, why repeating it?

wzzrd replied:
I didn't read your code well enough :) Will change, new pull request to follow

evgeni commented on bootstrap.py:
```diff
@@ -918,6 +929,44 @@ class MEOptions:
+        print_running("Stopping the Puppet agent for configuration update")
+        exec_failok("/sbin/service puppet stop")   # failok because people might be running Puppet from cron
+
+        print_running("Updating Puppet configuration")
```
I wonder if this all should be just

```
clean_puppet()
install_puppet_agent()
```

wzzrd replied:
No. That nukes puppet.conf, which might be bad for some users.

evgeni replied:
Yupp. It would be at least consistent with a fresh bootstrap, but I understand your concern.

evgeni commented on bootstrap.py:
```diff
@@ -765,6 +772,8 @@ class MEOptions:
         options.skip.append('puppet')
     if not options.removepkgs:
         options.skip.append('remove-obsolete-packages')
+    if options.new_capsule:
```
This should be done after `get_bootstrap_rpm()` via `API_PORT = get_api_port()`. Also this makes moving *to* the internal capsules impossible.

wzzrd replied:
Does it show really badly that I don't do this for a living ;) Will change, new pull request to follow

Commit:
Dynamically get API_PORT instead of hard-code it

Commit:
Put sys.exit(1) back in place

Was removed by accident

Commit:
Refactoring certificate installatin

Replaced 'manual' code with get_bootstrap_rpm() function. Requires to set
options.force to True to make get_bootstrap_rpm() to actually run
clean_katello_agent().

Commit:
Replace rhsmcertd restart code with functioncall

Don't 'manually' restart rhsmcertd, call enable_rhsmcertd instead. Also,
restarting goferd is already done elsewhere.

Commit:
Make puppet config change code RHEL5 safe

RHEL5's Puppet doesn't support 'puppet config set', so use ConfigParser
instead.

Commit:
Replace ConfigParser with callout to sed

I know it's not pretty, but it works on RHELs all over the place, it's
simple, it doesn't nuke /etc/puppet/puppet.conf and it's not based on
ConfigParser, which pukes on puppet.conf :/

Commit:
Do not run the prereqs bits if moving capsules

If moving a system to another capsule, do not run the RHEL5 prereqs
bits. When moving a system, assume the old situation already worked.

Commit:
Also change ca_server in stock RHEL5 puppet.conf

In the puppet.conf we deploy from Satellite 6, we write a ca_server
line. This line needs to be changed in puppet.conf as well when moving
capsules, else the CSR will end up on the Satellite and that won't work
for obvious reasons.

evgeni commented on bootstrap.py:
```diff
@@ -918,6 +928,46 @@ class MEOptions:
+        # > Puppet, OpenSCAP and update Puppet configuration
+        # > MANUAL SIGNING OF CSR OR MANUALLY CREATING AUTO-SIGN RULE STILL REQUIRED!
+        # > API doesn't have a public provision for creating auto-sign entries yet!
+        options.force = True  # Need to set this to make get_bootstrap_rpm() to perform the right way
```
can we change `get_bootstrap_rpm` to accept a `clean` Boolean, default it to `False` but pass `options.force` every time we call it instead. And then here don't set force to `true` but call `get_bootstrap_rpm(clean=True)`?

Just want to make sure that in future we don't actually want to omit something on `force=False` when moving capsules and can't or forget about it.

wzzrd replied:
So we could do this in two ways:

1. make get_bootstrap_rpm run clean_katello_agent() when either options.force is True *or* clean is True
2. change the code to always call get_bootstrap_rpm() like get_bootstrap_rpm(clean=options.force)

Preferences?

evgeni replied:
My initial thoughts were along the lines of 2, as it is less code.

@sideangleside what about you?

wzzrd replied:
Went with option 2 for now

evgeni commented on bootstrap.py:
```diff
@@ -918,6 +928,46 @@ class MEOptions:
+        update_host_capsule_mapping(options.foreman_fqdn, API_PORT, "content_source_id", capsule_id, host_id)
+        update_host_capsule_mapping(options.foreman_fqdn, API_PORT, "openscap_proxy_id", capsule_id, host_id)
+
+        print_running("Restarting rhsmcertd")
```
I think this should be "Enabling and restarting rhsmcertd" and be moved *inside* `enable_rhsmcertd` so it gets printed on the other parts too.

wzzrd replied:
Makes sense. Done.

evgeni commented on bootstrap.py:
```diff
@@ -459,9 +459,15 @@ def delete_json(url):
+    return call_api(url, data=jdata, method='PUT')
+
+
+def update_host_capsule_mapping(server, port, attribute, capsule_id, host_id):
```
We usually don't have `server` and `port` in the other functions, and just default to the global stuff. While not the nicest way, it would be consistent. I am torn.

wzzrd replied:
Can live with that. Am updating some code now, will check whether this works later this week and also include the switch back to built-in proxy as test case.

Commit:
Make update_host_capsule_mapping use global vars

Not pretty, but consistent, at least according to Zhenech ;)

Commit:
Do not force options.force, but parameterize

Don't force override a user provided option, instead add a parameter to
get_bootstrap_rpm.

Commit:
Change get_bootstrap_rpm to accept clean parameter

This allows the get_bootstrap_rpm to accept a clean parameter, which
would usually be equal to options.force, but can be overridden for some
use cases.

Commit:
Move status message into enable_rhsmcertd

Get status messages everytime you run this method for free!

Commit:
Make travis happy

E712: comparison to True should be 'if cond is True' or 'if cond'

evgeni commented on bootstrap.py:
```diff
@@ -160,13 +160,13 @@ def install_prereqs():
     """
     Retrieve Client CA Certificate RPMs from the Satellite 6 server.
     Uses --insecure options to curl(1) if instructed to download via HTTPS
     If called with --force, calls clean_katello_agent().
```
this comment should be updated to match the new `clean` param

wzzrd replied:
Will add.

evgeni commented on bootstrap.py:
```diff
@@ -918,11 +929,49 @@ class MEOptions:
+        API_PORT = get_api_port()
+
+        print_running("Calling Foreman API to update content source, puppet master, puppet ca and openscap proxy records for %s" % FQDN)
+        capsule_id = return_matching_katello_key('capsules', 'name="%s"' % options.foreman_fqdn, 'id', False)
```
any reason not to hit the Foreman API for that?

```
capsule_id = return_matching_foreman_key('smart_proxies', 'name="%s"' % options.foreman_fqdn, 'id', False)
```

wzzrd replied:
Not really (I think). Just picked the first one in the API docs that fitted my requirements. Would have to test whether or not the Foreman API works as well (but would guess so). Any reason as why this would matter?

evgeni replied:
This would make (this part of the code) also work on Foreman installs without Katello. Plus I'd hope we can someday get rid of the Katello-specific API completely.

I gave the APIs on 1.15 upstream nightly and 6.2.8 Satellite a shot yesteday, the results looked identical enough for your needs.

Commit:
Alter docstring for get_bootstrap_rpm()

Now reflects new clean parameter

Commit:
Add some documentation to README.md

Add a description of the new --new-capsule functionality

Commit:
Calling foreman api instead of katello

For future-proofness, and foreman api should work just as well as
katello api for this use case, according to Zhenech.

evgeni commented on README.md:
```diff
@@ -170,6 +184,18 @@ By default, bootstrap.py does not delete the system's profile from the legacy pl
+
+
+~~~
+# ./bootstrap.py --new-capsule \
```
this needs credentials, no?

evgeni commented on README.md:
```diff
@@ -7,6 +7,9 @@ bootstrap Script for migrating existing running systems to Foreman with the Kate
 Network Classic and get it registered to Foreman & Katello.
 
+Optionally, you can also move systems from one Satellite 6 system or Capsule to
+another by using the --new-capsule option.
```
`--new-capsule` so that it properly formated

evgeni commented on bootstrap.py:
```diff
@@ -918,11 +930,49 @@ class MEOptions:
+
+        enable_rhsmcertd()
+
+        print_running("Stopping the Puppet agent for configuration update")
```
this should be wrapped in `if 'puppet' not in options.skip`?

wzzrd replied:
OK

evgeni commented on bootstrap.py:
```diff
@@ -918,11 +930,49 @@ class MEOptions:
+        install_katello_agent()
+        API_PORT = get_api_port()
+
+        print_running("Calling Foreman API to update content source, puppet master, puppet ca and openscap proxy records for %s" % FQDN)
```
this should be wrapped in `if 'foreman' not in options.skip`?

wzzrd replied:
The Puppet stuff only, right? Not the content source? What about the OpenSCAP proxy?

Should I wrap the puppet_proxy_id and puppet_ca_proxy_id lines in `if 'puppet' not in options.skip`, too?

sideangleside commented on bootstrap.py:
```diff
@@ -918,11 +930,49 @@ class MEOptions:
+        update_host_capsule_mapping("puppet_proxy_id", capsule_id, host_id)
+        update_host_capsule_mapping("puppet_ca_proxy_id", capsule_id, host_id)
+        update_host_capsule_mapping("content_source_id", capsule_id, host_id)
+        update_host_capsule_mapping("openscap_proxy_id", capsule_id, host_id)
```
It is not guaranteed that every capsule runs the OpenSCAP feature. In testing, migrating from one Capsule to another (where the destination did not have the OpenSCAP feature), bootstrap.py fails with an error message similar to

~~~
RUNNING], [2017-03-30 06:22:12], [Calling Foreman API to update content source, puppet master, puppet ca and openscap proxy records for devnode-115.example.com] 
An error occured: HTTP Error 422: Unprocessable Entity
url: https://capsule-2.example.com:8443/api/v2/hosts/231
code: 422
data: {
  "host": {
    "openscap_proxy_id": "3"
  }
}
error: {
  "error": {
    "errors": {
      "openscap_proxy_id": [
        "must have Openscap feature"
      ]
    }, 
    "id": 231, 
    "full_messages": [
      "Openscap Capsule must have Openscap feature"
    ]
  }
}

~~~

It may make sense to check the Capsule's features, and then update the host if (and only if) it has the OpenSCAP feature.

wzzrd replied:
Fair point, will look into that.

sideangleside commented on bootstrap.py:
```diff
@@ -918,11 +930,49 @@ class MEOptions:
+        delete_file("/var/lib/puppet/client_data/catalog/%s.json" % FQDN)
+
+        print_running("Starting Puppet and waiting for cert; please go and sign it now!")
+        exec_failok("/usr/bin/puppet agent --noop -t --waitforcert 15")
```
For consistency would it make sense to update this line to match what is in the `install_puppet_agent()` function: 

~~~
    print_generic("Running Puppet in noop mode to generate SSL certs")
    print_generic("Visit the UI and approve this certificate via Infrastructure->Capsules")
    print_generic("if auto-signing is disabled")
    exec_failexit("/usr/bin/puppet agent --test --noop --tags no_such_tag --waitforcert 10")
~~~

wzzrd replied:
Dito

ekohl commented on bootstrap.py:
```diff
@@ -459,9 +461,15 @@ def delete_json(url):
+
+def update_host_capsule_mapping(attribute, capsule_id, host_id):
+    url = "https://" + options.foreman_fqdn + ":" + str(API_PORT) + "/api/v2/hosts/" + str(host_id)
+    jdata = json.loads('{"host": {"%s": "%s"}}' % (attribute, capsule_id))
```
I believe this line is equal to ```jdata = {'host': {attribute: capsule_id}}```
